### PR TITLE
New Feature: Level Caps

### PR DESCRIPTION
**Pull request by AlexOn1ine**
## Level Caps 
**Hard Caps**
 - No explanation need. Once you reach a certain threshold you can't go above it.

**Soft Caps**
 - Once cap is reached you still get exp with diminishing returns. Some defaults are given but it's fully customizable.

**Rare Candy toggle** 
 - Rare Candies allow to bypass the level cap unless `B_RARE_CANDY_CAP` is set to true.

**Level Cap type** 
 - Level caps can be set with a flag or variable. An example is given for flags with Emerald level caps. For the variable an event var is needed. 

Credits to SBird for initial outline and PokemonCrazy for soft caps, though I deviated from their implementation.

### CODE REVIEW

**Comment by PCG06** [comment]
Can you try an interaction with EXP candies please?

**Comment by AlexOn1ine** [comment]
> Can you try an interaction with EXP candies please?

Can't be used if `B_RARE_CANDY_CAP` is set.

File: include/level_caps.h [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+#if B_EXP_CAP_TYPE == EXP_CAP_HARD || B_EXP_CAP_TYPE == EXP_CAP_SOFT
+#if B_LEVEL_CAP_TYPE != LEVEL_CAP_FLAG_LIST && B_LEVEL_CAP_TYPE != LEVEL_CAP_VARIABLE
+#error "Invalid choice for B_LEVEL_CAP_TYPE, must be of [LEVEL_CAP_FLAG_LIST, LEVEL_CAP_VARIABLE]"
+#endif
+#if B_LEVEL_CAP_TYPE == LEVEL_CAP_VARIABLE && B_LEVEL_CAP_VARIABLE == 0
+#error "B_LEVEL_CAP_TYPE set to LEVEL_CAP_VARIABLE, but no variable chosen for B_LEVEL_CAP_VARIABLE, set B_LEVEL_CAP_VARIABLE to a valid event variable"
+#endif
+#endif
+
+u8 GetCurrentLevelCap();
```
Comment (DizzyEggg) [inline include/level_caps.h:34]: Let's use u32 for all arguments.
Also if the function takes no arguments, use `(void)`

File: src/level_caps.c [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    {FLAG_BADGE05_GET, 31},
+    {FLAG_BADGE06_GET, 33},
+    {FLAG_BADGE07_GET, 42},
+    {FLAG_BADGE08_GET, 46},
+    {FLAG_IS_CHAMPION, 58},
+};
+
+u8 GetCurrentLevelCap(void)
+{
+    u8 i;
```
Comment (DizzyEggg) [inline src/level_caps.c:22]: u32 i;

File: src/level_caps.c [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    }
+    else if (B_LEVEL_CAP_TYPE == LEVEL_CAP_VARIABLE)
+    {
+        return VarGet(B_LEVEL_CAP_VARIABLE);
+    }
+
+    return MAX_LEVEL;
+}
+
+const u16 sExpScalingDown[5] = { 4, 8, 16, 32, 64 };
```
Comment (DizzyEggg) [inline src/level_caps.c:40]: const u8 to reduce rom size

**Review by DizzyEggg — CHANGES_REQUESTED** [review]

**Comment by AlexOn1ine** [comment]
Reviews applied :eyes:

**Comment by katykat5099** [comment]
@AlexOn1ine there are level caps in thexamans challenges.

**Comment by AlexOn1ine** [comment]
> @AlexOn1ine there are level caps in thexamans challenges.

I haven't looked into thexamans version level caps but this one extensively covers all cases that users can easily choose from.

**Review by Bassoonian — APPROVED** [review]

**Comment by zsd7200** [comment]
hey, PokemonCrazy here! thank you for putting this in! just saw this in the release and freaked out seeing my name in the release notes lol

i was just getting back into my project after like 3 years of being dormant and planned to clean this code up a bit but now i can just use this, this is awesome :)

**Pull request merged**